service/history/hsm/tree.go

```diff
@@ -280,17 +284,25 @@ func (n *Node) AddChild(key Key, data any) (*Node, error) {
 	if err != nil {
 		return nil, err
 	}
+
+	nextVersionedTransition := &persistencespb.VersionedTransition{
+		NamespaceFailoverVersion: n.backend.GetCurrentVersion(),
+		// The transition count for the backend is only incremented when closing the current transaction,
+		// but any change to state machine node is a state transtion,
+		// so we can safely using next transition count here is safe.
+		TransitionCount: n.backend.NextTransitionCount(),
+	}
 	node := &Node{
 		Key:        key,
 		Parent:     n,
 		definition: def,
 		registry:   n.registry,
 		persistence: &persistencespb.StateMachineNode{
-			Children:                              make(map[string]*persistencespb.StateMachineMap),
-			Data:                                  serialized,
-			InitialNamespaceFailoverVersion:       n.backend.GetCurrentVersion(),
-			InitialMutableStateTransitionCount:    n.backend.TransitionCount(),
-			LastUpdateMutableStateTransitionCount: n.backend.TransitionCount(),
+			Children:                      make(map[string]*persistencespb.StateMachineMap),
+			Data:                          serialized,
+			InitialVersionedTransition:    nextVersionedTransition,
+			LastUpdateVersionedTransition: nextVersionedTransition,
+			TransitionCount:               0,
```

Comment:
Shall we initialize it to 1? Or we will always do a MachineTransition for a node after adding it?

Reply:
I think you're right, better to initialize to 1.

Reply:
Hmm... we can't ever generate tasks for a machine without going through `MachineTransition` so it's probably not super critical what we initialize this to but having the initial value be 1 is less error prone and makes less assumptions.

Reply:
Synced offline. Not too much difference between 0 and 1. Will keep as 0 for now.